pkg/db/db.go

```diff
@@ -1,268 +1,183 @@
-	StoreMetadata(metadata db.Metadata, dir string) (err error)
+// WithClock takes a clock
+func WithClock(clock clock.Clock) Option {
+	return func(opts *options) {
+		opts.clock = clock
+	}
 }
 
 // Client implements DB operations
 type Client struct {
-	dbc          dbOperation
-	githubClient github.Operation
-	pb           indicator.ProgressBar
-	clock        clock.Clock
-	metadata     Metadata
+	*options
+
+	cacheDir string
+	metadata metadata.Client
+	quiet    bool
 }
 
 // NewClient is the factory method for DB client
-func NewClient(dbc dbOperation, githubClient github.Operation, pb indicator.ProgressBar, clock clock.Clock, metadata Metadata) Client {
-	return Client{
-		dbc:          dbc,
-		githubClient: githubClient,
-		pb:           pb,
-		clock:        clock,
-		metadata:     metadata,
+func NewClient(cacheDir string, quiet bool, opts ...Option) *Client {
+	o := &options{
+		clock: clock.RealClock{},
 	}
-}
 
-// NeedsUpdate check is DB needs update
-func (c Client) NeedsUpdate(cliVersion string, light, skip bool) (bool, error) {
-	dbType := db.TypeFull
-	if light {
-		dbType = db.TypeLight
+	for _, opt := range opts {
+		opt(o)
+	}
+
+	return &Client{
+		options:  o,
+		cacheDir: cacheDir,
+		metadata: metadata.NewClient(cacheDir),
+		quiet:    quiet,
 	}
+}
 
-	metadata, err := c.metadata.Get()
+// NeedsUpdate check is DB needs update
+func (c *Client) NeedsUpdate(cliVersion string, skip bool) (bool, error) {
+	meta, err := c.metadata.Get()
 	if err != nil {
 		log.Logger.Debugf("There is no valid metadata file: %s", err)
 		if skip {
 			log.Logger.Error("The first run cannot skip downloading DB")
 			return false, xerrors.New("--skip-update cannot be specified on the first run")
 		}
-		metadata = db.Metadata{} // suppress a warning
+		meta = metadata.Metadata{Version: db.SchemaVersion}
 	}
 
-	if db.SchemaVersion < metadata.Version {
+	if db.SchemaVersion < meta.Version {
 		log.Logger.Errorf("Trivy version (%s) is old. Update to the latest version.", cliVersion)
 		return false, xerrors.Errorf("the version of DB schema doesn't match. Local DB: %d, Expected: %d",
-			metadata.Version, db.SchemaVersion)
+			meta.Version, db.SchemaVersion)
 	}
 
 	if skip {
-		if err = c.validate(dbType, metadata); err != nil {
-			return false, err
+		if err = c.validate(meta); err != nil {
+			return false, xerrors.Errorf("validate error: %w", err)
 		}
 		return false, nil
 	}
 
-	if db.SchemaVersion != metadata.Version || metadata.Type != dbType {
+	if db.SchemaVersion != meta.Version {
 		return true, nil
 	}
 
-	return !c.isNewDB(metadata), nil
+	return !c.isNewDB(meta), nil
 }
 
-func (c Client) validate(dbType db.Type, metadata db.Metadata) error {
-	if db.SchemaVersion != metadata.Version {
+func (c *Client) validate(meta metadata.Metadata) error {
+	if db.SchemaVersion != meta.Version {
 		log.Logger.Error("The local DB is old and needs to be updated")
 		return xerrors.New("--skip-update cannot be specified with the old DB")
```

Comment:
I think this is ok long term. But for now it will break processing for everyone who is using `--skip-update`  option (until they update db using `oras`)

Reply:
We added breaking change to Trivy DB, so the newer Trivy doesn't work with an old database. `--skip-update` should not work with the old DB. In other words, the following case should fail:

- Trivy: v0.24.0
- Trivy DB: v1

Reply:
Should it be documented somewhere?

Reply:
Is the error not enough? If you feel so, we can document it.

Reply:
I checked and seems there is no good place to document it. Change log has reference to this PR. And it's good enough
But I would change message to mention "schema". Like "Local DB has schema version which is not supported anymore.  It needs to be updated". Just to distigush that it's not regular db update is required.

Reply:
Thanks. I've updated. Also, I mentioned it in the trouble shooting doc.
https://github.com/aquasecurity/trivy/pull/1539/commits/576a35457926692cc7530d33658d2453125d6be8